Implement PnPjs and Fluent UI

andreasim24 opened the pull request:
- Convert React Class Component to Function Component
- Convert SPHttps to PnPjs
- Implement Fluent UI Component

andreasim24 committed:
Implement PnPjs and Fluent UI

- Convert React Class Component to Function Component
- Convert SPHttps to PnPjs
- Implement Fluent UI Component

AndreyButyrchik commented on src/pnpjsConfig.ts:
```diff
@@ -0,0 +1,17 @@
+import "@pnp/sp/items";
+import "@pnp/sp/batching";
+
+let _sp: SPFI = null;
```
Not the best way to singleton implementation. I suggest using class with own `_sp` property instead of global variable, because this implementation can cause issues in sharepoint.
spfx initialization depends on context. In some cases when you navigate between pages in sharepoint it reloads page partially (not full reload) and in that case context will be changed, but global variable still will be initialized.

Here is an example how to implement singleton. Make initialization in onInit method of ToDoListWebPart.ts - https://refactoring.guru/design-patterns/singleton/typescript/example#:~:text=Singleton%20in%20TypeScript%20Singleton%20is%20a%20creational%20design,the%20same%20pros%20and%20cons%20as%20global%20variables.
Also you can read about partial reload - https://www.eliostruyf.com/navigating-to-other-pages-in-sharepoint-framework-from-code

AndreyButyrchik commented on src/webparts/todolist/components/ToDoList.module.scss:
```diff
@@ -10,86 +10,15 @@
-.itemIcon {
-  margin-left: 4px;
+.statusPending {
+  background-color: "#ffe092";
```
I would prefer to use on of the accent colors intead own one.
https://developer.microsoft.com/en-us/fluentui#/styles/web/colors/theme-slots

So, your color will be smth like `color: "[theme: yellow, default:#ffe092]"`

AndreyButyrchik commented on src/webparts/todolist/components/ToDoList.tsx:
```diff
@@ -1,129 +1,213 @@
+              <Stack
+                horizontalAlign="center"
+                verticalAlign="center"
+                className={mergeStyles({
```
I suggest doing it using css classes from scss file other than this way, or at least move all of these blocks of code to separate file.

AndreyButyrchik commented on src/webparts/todolist/components/ToDoList.tsx:
```diff
@@ -1,129 +1,213 @@
-  public componentDidMount(): void {
-    this._getListData();
-  }
+  const getListData = async (): Promise<void> => {
```
You need to initialize sp pnp in ToDoListWebPart class, at onInit method

AndreyButyrchik commented on src/webparts/todolist/components/ToDoList.tsx:
```diff
@@ -1,129 +1,213 @@
-    this._getListData();
-  }
+  const getListData = async (): Promise<void> => {
+    const LIST_NAME = "To do list";
```
List name should't be hardcoded, put it as an parameter to webpart properties and add ability to change it via property pane

AndreyButyrchik commented on src/webparts/todolist/components/ToDoList.tsx:
```diff
@@ -1,129 +1,213 @@
-        </h2>
+  React.useEffect(() => {
+    getListData();
+  }, []);
```
should depends on list item name

AndreyButyrchik commented on src/webparts/todolist/components/ToDoList.tsx:
```diff
@@ -1,129 +1,213 @@
+  padding: 10
+};
+
+const ItemList: React.FC<IItemListProps> = (props: IItemListProps) => {
```
Put this functional component in separate file

AndreyButyrchik commented on src/interfaces.ts:
```diff
@@ -0,0 +1,31 @@
+}
+
+export interface IItemListProps {
+  spHttpClient: SPHttpClient;
```
spHttpClient not uses in component, need to delete it

AndreyButyrchik commented on src/interfaces.ts:
```diff
@@ -0,0 +1,31 @@
+
+export interface IItemListProps {
+  spHttpClient: SPHttpClient;
+  webUrl: string;
```
same as spHttpClient

AndreyButyrchik commented on src/webparts/todolist/components/ToDoList.tsx:
```diff
@@ -1,129 +1,213 @@
         <ItemList
-          spHttpClient={this.props.spHttpClient}
-          webUrl={this.props.websiteUrl}
+          spHttpClient={spHttpClient}
```
spHttpClient and webUrl are not use in ItemListComponent

AndreyButyrchik commented on src/webparts/todolist/ToDoListWebPart.ts:
```diff
@@ -29,7 +29,8 @@ export default class ToDoListWebPart extends BaseClientSideWebPart<IToDoListWebP
         isDarkTheme: this._isDarkTheme,
         environmentMessage: this._environmentMessage,
         hasTeamsContext: !!this.context.sdks.microsoftTeams,
```
hasTeamsContext, environmentMessage, isDarkTheme are not use in component. Need to delete this properties

AndreyButyrchik commented on src/interfaces.ts:
```diff
@@ -0,0 +1,31 @@
+  errorMessage: any;
+}
+
+export interface IToDoListProps {
```
Delete unused properties from interface

andreasim24 committed:
fix remarks task 2

- change pnpjs initialization method
- use fluent ui theme colors
- use scss class for list items status component
- add propertypane dropdown to choose lists (dynamically change list name)
- remove unused props and interfaces

AndreyButyrchik commented on src/webparts/todolist/components/ToDoList.module.scss:
```diff
@@ -19,6 +9,21 @@
-  background-color: "#ffe092";
-  color: "#997825";
+  background-color: "[theme: warning, default:#fff4ce]";
+  color: "[theme: OrangeYellow20, default:#986f0b]";
```
Not sure that OrangeYellow20 is correct token name. Most popular colors you can find here.
https://developer.microsoft.com/en-us/fluentui#/styles/web/colors/theme-slots

<img width="723" alt="image" src="https://user-images.githubusercontent.com/32543592/204173014-be557d76-6543-48ce-97b6-40f083ffe1c0.png">

Also you can open console at any sharepoint site and invoke this expression `__themeState__.theme` then you will get theme object

AndreyButyrchik commented on src/webparts/todolist/components/ToDoList.tsx:
```diff
@@ -167,7 +115,7 @@ const ItemList: React.FC<IItemListProps> = (props: IItemListProps) => {
 
   React.useEffect(() => {
     getListData();
```
Will be better to put props.listName as a parameter in method. Let's make function more generic

AndreyButyrchik commented on src/webparts/todolist/ToDoListWebPart.ts:
```diff
@@ -3,106 +3,44 @@ import * as ReactDom from "react-dom";
-    });
+  protected async onInit(): Promise<void> {
+    await super.onInit();
+    this._sp = spfi().using(SPFx(this.context));
```
Will be better to make it using service like here https://dev.azure.com/engagesq/_git/Engage%20Squared%20-%20Intranet%20Components?path=/components/content-author-dashboard/src/services/ListService.ts. But for small webparts this initialization also fine

andreasim24 committed:
fix remarks task 2

andreasim24 committed:
fiix last remarks
